### PR TITLE
Add a warning to File about mutability.

### LINKED ISSUE #47708: File::set_len uses an immutable reference to self
I am creating a wrapper for the File type and came across an issue that I thought was odd.

- [File::set_len](https://doc.rust-lang.org/std/fs/struct.File.html#method.set_len)
- [File::set_permisions](https://doc.rust-lang.org/std/fs/struct.File.html#method.set_permisions)

Both of these methods take an _immutable_ reference (`&self`) to the File instead of a _mutable_ reference (`&mut self`).

I'm not so much concerned about _memory safety_ per se, as I assume that the operating system guarantees that operaions are memory safe, it is just that this is extremely unexpected. If I pass a function an immutable reference to a `File`, I wouldn't expect it to be able to truncate the file!

Obviously this is a breaking change so probably can't be changed -- however, I was thinking we could at least add docs explaining why the reference is immutable.

### PR DESCRIPTION
**Pull request by alercah**
Fixes #47708.

### CODE REVIEW

File: src/libstd/fs.rs [hunk 1]
```diff
@@ -81,9 +81,18 @@ use time::SystemTime;
 /// # }
 /// ```
 ///
+/// Note that, although read and write methods require a `&mut File`, because
+/// of the interfaces for [`Read`] and [`Write`], it is still possible to
+/// modify a file through a `&File`, either through methods that take `&File`
+/// or by retrieving a raw OS filehandle and modifying the file that way.
+/// Additionally, many operating systems allow concurrent modification of files
+/// by different processes. Care should be taken not to assume that holding a
+/// `&File` means that the file will not change.
```
Comment (joshtriplett) [inline src/libstd/fs.rs]: This is a bit of a nitpick, but could you please phrase this in active voice?

"..., the holder of a `&File` can still modify that file, either through methods ..."

"... Avoid assuming that holding a `&File` means ..."

**Comment by sfackler** [comment]
LGTM, but letting one of the docs folks r+.

File: src/libstd/fs.rs [hunk 2]
```diff
@@ -81,9 +81,18 @@ use time::SystemTime;
 /// # }
 /// ```
 ///
+/// Note that, although read and write methods require a `&mut File`, because
+/// of the interfaces for [`Read`] and [`Write`], it is still possible to
+/// modify a file through a `&File`, either through methods that take `&File`
+/// or by retrieving a raw OS filehandle and modifying the file that way.
```
Comment (sfackler) [inline src/libstd/fs.rs]: "file handle"

**Review by joshtriplett — CHANGES_REQUESTED** [review]
This looks great to me, and I'm glad to see these kinds of clarifying comments that help people understand the high-level semantics around what they're trying to do.

I posted one note about grammar, but otherwise I'd love to see this merged. Thanks!

File: src/libstd/fs.rs (hunk 2 above)
Comment (joshtriplett) [reply src/libstd/fs.rs]: Or "underlying OS object", since on Linux it'd be a file *descriptor*.

**Comment by alercah** [comment]
Comments fixed.

**Review by joshtriplett — APPROVED** [review]

**Pull request merged**